tools/bioimaging/main.py

```diff
@@ -0,0 +1,78 @@
+"""
+Predict images using AI models from BioImage
+"""
+
+import argparse
+
+import imageio
+import numpy as np
+import torch
+
+
+def find_dim_order(user_in_shape, input_image):
+    image_shape = list(input_image.shape)
+    correct_order = user_in_shape.split(",")[::-1]
+    correct_order = [int(i) for i in correct_order if i != "1"]
+    if (correct_order[0] == image_shape[-1]) and (correct_order != image_shape):
+        input_image = torch.tensor(input_image.transpose())
+    return input_image, correct_order
+
+
+if __name__ == "__main__":
+    arg_parser = argparse.ArgumentParser()
+    arg_parser.add_argument("-im", "--imaging_model", required=True, help="Input BioImage model")
+    arg_parser.add_argument("-ii", "--image_file", required=True, help="Input image file")
+    arg_parser.add_argument("-is", "--image_size", required=True, help="Input image file's size")
+
+    # get argument values
+    args = vars(arg_parser.parse_args())
+    model_path = args["imaging_model"]
+    input_image_path = args["image_file"]
+
+    # load all embedded images in TIF file
+    test_data = imageio.v3.imread(input_image_path, index="...")
+    test_data = np.squeeze(test_data)
+    test_data = test_data.astype(np.float32)
+
+    # assess the correct dimensions of TIF input image
+    input_image_shape = args["image_size"]
+    im_test_data, shape_vals = find_dim_order(input_image_shape, test_data)
+
+    # load model
+    model = torch.load(model_path)
+    model.eval()
+
+    # find the number of dimensions required by the model
+    target_dimension = 0
+    for param in model.named_parameters():
+        target_dimension = len(param[1].shape)
+        break
+    current_dimension = len(list(im_test_data.shape))
+
+    # update the dimensions of input image if the required image by
+    # the model is smaller
+    slices = tuple(slice(0, s_val) for s_val in shape_vals)
+
+    # apply the slices to the reshaped_input
+    im_test_data = im_test_data[slices]
+    exp_test_data = torch.tensor(im_test_data)
+
+    # expand input image's dimensions
+    for i in range(target_dimension - current_dimension):
+        exp_test_data = torch.unsqueeze(exp_test_data, i)
+
+    # make prediction
+    pred_data = model(exp_test_data)
+    pred_data_output = pred_data.detach().numpy()
+
+    # save original image matrix
+    np.save("output_predicted_image_matrix.npy", pred_data_output)
+
+    # post process predicted file to correctly save as TIF file
+    pred_data = torch.squeeze(pred_data)
+    pred_numpy = pred_data.detach().numpy()
+    pred_numpy = pred_numpy * 255
+    pred_numpy = pred_numpy.astype(np.uint8)
```

Comment:
the original predicted values are floats (< 1). When such a matrix is saved as a TIF file, it does not show anything after opening (normalised).

![output_empty](https://github.com/user-attachments/assets/18923a02-fe33-4f25-b21d-936773950d5c)

This issue I have seen in the `sample_output.tif` files for a few BioImage models.

![empty_tiff](https://github.com/user-attachments/assets/8f9289fe-718a-4d24-a34f-498771010f51)

To avoid it, I multiply it by 255 and save it as a matrix of Int

Reply:
I'm pretty sure that this is an issue with your image viewer:

> In [1]: import imageio
> 
> In [2]: import numpy as np
> 
> In [3]: im1 = np.random.rand(50,50)
> 
> In [4]: im1.min(), im1.max()
> Out[4]: (0.000287018172377973, 0.999300887698757)
> 
> In [5]: imageio.v3.imwrite('test.tif', im1)
> 
> In [6]: im2 = imageio.v3.imread('test.tif')
> 
> In [7]: np.allclose(im1, im2)
> Out[7]: True